File: src/lib/historical-data-normalizer.ts
import type { Candle } from '@/types'

export type HistoricalNormalizationStatus = 'OK' | 'WARNING' | 'FAILED'
export type AdjustedStatus = 'adjusted' | 'unadjusted' | 'unknown'

export interface HistoricalDataNormalizerResult {
  symbol: string
  provider: string
  normalized_dataset: Candle[]
  row_count: number
  timezone: string
  currency?: string
  adjusted_status: AdjustedStatus
  normalization_status: HistoricalNormalizationStatus
  issues: { code: string; message: string; details?: Record<string, unknown> }[]
  warnings: { code: string; message: string; details?: Record<string, unknown> }[]
  raw_diagnostics: Record<string, unknown>
}

const PRICE_KEYS = ['open', 'high', 'low', 'close'] as const
const REQUIRED_KEYS = [...PRICE_KEYS, 'volume'] as const

function normalizeKey(key: string) {
  const normalized = key.trim().toLowerCase().replace(/[\s.-]+/g, '_')
  if (['date', 'datetime', 'timestamp', 'time'].includes(normalized)) return 'time'
  if (['adjclose', 'adj_close', 'adjusted_close'].includes(normalized)) return 'adj_close'
  return normalized
}

function numberOrNull(value: unknown) {
  const numeric = Number(value)
  return Number.isFinite(numeric) ? numeric : null
}

function parseTime(value: unknown) {
  if (value instanceof Date) return Math.floor(value.getTime() / 1000)
  if (typeof value === 'number' && Number.isFinite(value)) {
    return value > 10_000_000_000 ? Math.floor(value / 1000) : Math.floor(value)
  }
  if (typeof value === 'string') {
    const numeric = Number(value)
    if (Number.isFinite(numeric)) return numeric > 10_000_000_000 ? Math.floor(numeric / 1000) : Math.floor(numeric)
    const parsed = Date.parse(value)
    return Number.isFinite(parsed) ? Math.floor(parsed / 1000) : null
  }
  return null
}

function adjustedStatus(metadata: Record<string, unknown>): AdjustedStatus {
  for (const key of ['adjusted', 'is_adjusted', 'auto_adjust', 'uses_adjusted_close']) {
    const value = metadata[key]
    if (typeof value === 'boolean') return value ? 'adjusted' : 'unadjusted'
    if (typeof value === 'string') {
      const normalized = value.toLowerCase()
      if (['true', 'yes', '1', 'adjusted'].includes(normalized)) return 'adjusted'
      if (['false', 'no', '0', 'raw', 'unadjusted'].includes(normalized)) return 'unadjusted'
    }
  }
  return 'unknown'
}

export function normalizeHistoricalData(input: {
  symbol: string
  provider: string
  market: string
  timeframe: string
  raw_dataset: Record<string, unknown>[]
  metadata?: Record<string, unknown>
}): HistoricalDataNormalizerResult {
  const metadata = input.metadata ?? {}
  const issues: HistoricalDataNormalizerResult['issues'] = []
  const warnings: HistoricalDataNormalizerResult['warnings'] = []
  const normalizedRows = input.raw_dataset.map((row) => {
    const normalized: Record<string, unknown> = {}
    for (const [key, value] of Object.entries(row)) normalized[normalizeKey(key)] = value
    return normalized
  })

  const sourceKeys = new Set(normalizedRows.flatMap(row => Object.keys(row)))
  const missingColumns = REQUIRED_KEYS.filter(key => !sourceKeys.has(key))
  if (!sourceKeys.has('time')) issues.push({ code: 'MISSING_TIME_COLUMN', message: 'Raw dataset has no date/time column.' })
  if (missingColumns.length) issues.push({ code: 'MISSING_OHLCV_COLUMNS', message: 'Raw dataset is missing OHLCV columns.', details: { missing_columns: missingColumns } })

  if (issues.length) {
    return {
      symbol: input.symbol,
      provider: input.provider,
      normalized_dataset: [],
      row_count: 0,
      timezone: String(metadata.timezone ?? metadata.tz ?? 'unknown'),
      currency: typeof metadata.currency === 'string' ? metadata.currency : undefined,
      adjusted_status: adjustedStatus(metadata),
      normalization_status: 'FAILED',
      issues,
      warnings,
      raw_diagnostics: { input_rows: input.raw_dataset.length, output_rows: 0 },
    }
  }

  const dedupe = new Set<string>()
  const candles: Candle[] = []
  let droppedRows = 0
  let duplicateRows = 0
  for (const row of normalizedRows) {
    const candle = {
      time: parseTime(row.time),
      open: numberOrNull(row.open),
      high: numberOrNull(row.high),
      low: numberOrNull(row.low),
      close: numberOrNull(row.close),
      volume: numberOrNull(row.volume),
    }
    if (candle.time === null || REQUIRED_KEYS.some(key => candle[key] === null)) {
      droppedRows += 1
      continue
    }
    if (PRICE_KEYS.some(key => (candle[key] as number) <= 0)) {
      droppedRows += 1
      issues.push({ code: 'NON_POSITIVE_PRICES_DROPPED', message: 'Rows with zero or negative prices were rejected.' })
      continue
    }
    const finalCandle = candle as Candle
    const signature = JSON.stringify(finalCandle)
    if (dedupe.has(signature)) {
      duplicateRows += 1
      continue
    }
    dedupe.add(signature)
    candles.push(finalCandle)
  }

  candles.sort((a, b) => a.time - b.time)
  const zeroVolume = candles.filter(candle => candle.volume <= 0).length
  if (zeroVolume) warnings.push({ code: 'ZERO_OR_NULL_VOLUME', message: 'Dataset contains zero or null volume.', details: { rows: zeroVolume } })
  if (duplicateRows) warnings.push({ code: 'DUPLICATE_ROWS_REMOVED', message: 'Exact duplicate candles were removed.', details: { rows: duplicateRows } })
  if (sourceKeys.has('adj_close') && adjustedStatus(metadata) === 'unknown') {
    warnings.push({ code: 'ADJUSTMENT_AMBIGUOUS', message: 'adj_close exists but OHLC adjustment status is not declared.' })
  }

  const status: HistoricalNormalizationStatus = candles.length === 0 ? 'FAILED' : issues.length || warnings.length || droppedRows ? 'WARNING' : 'OK'
  return {
    symbol: input.symbol,
    provider: input.provider,
    normalized_dataset: candles,
    row_count: candles.length,
    timezone: String(metadata.timezone ?? metadata.tz ?? 'unknown'),
    currency: typeof metadata.currency === 'string' ? metadata.currency : undefined,
    adjusted_status: adjustedStatus(metadata),
    normalization_status: status,
    issues,
    warnings,
    raw_diagnostics: { input_rows: input.raw_dataset.length, output_rows: candles.length, dropped_rows: droppedRows, duplicate_rows_removed: duplicateRows },
  }
}

File: quant-engine/historical_data_normalizer.py
from __future__ import annotations

from typing import Any, Dict, List, Optional

import pandas as pd


PRICE_COLUMNS = ["open", "high", "low", "close"]
REQUIRED_COLUMNS = PRICE_COLUMNS + ["volume"]
TIME_ALIASES = {"date", "datetime", "timestamp", "time"}
COLUMN_ALIASES = {
    "date": "time",
    "datetime": "time",
    "timestamp": "time",
    "time": "time",
    "open": "open",
    "high": "high",
    "low": "low",
    "close": "close",
    "adjclose": "adj_close",
    "adj_close": "adj_close",
    "adj_close_": "adj_close",
    "adjusted_close": "adj_close",
    "volume": "volume",
}


def _issue(code: str, message: str, details: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    item: Dict[str, Any] = {"code": code, "message": message}
    if details:
        item["details"] = details
    return item


def _normalize_name(name: Any) -> str:
    text = str(name).strip().lower().replace(" ", "_").replace("-", "_")
    text = text.replace(".", "_")
    return COLUMN_ALIASES.get(text, text)


def _raw_to_frame(raw_dataset: Any) -> pd.DataFrame:
    if isinstance(raw_dataset, pd.DataFrame):
        frame = raw_dataset.copy()
    else:
        frame = pd.DataFrame(list(raw_dataset or []))

    if not isinstance(frame.index, pd.RangeIndex):
        index_name = frame.index.name or "time"
        if _normalize_name(index_name) in TIME_ALIASES or isinstance(frame.index, pd.DatetimeIndex):
            frame = frame.reset_index().rename(columns={index_name: "time", "index": "time"})

    rename_map = {column: _normalize_name(column) for column in frame.columns}
    return frame.rename(columns=rename_map)


def _metadata_adjusted_status(metadata: Dict[str, Any]) -> str:
    for key in ("adjusted", "is_adjusted", "auto_adjust", "uses_adjusted_close"):
        if key not in metadata:
            continue
        value = metadata.get(key)
        if isinstance(value, bool):
            return "adjusted" if value else "unadjusted"
        if isinstance(value, str):
            normalized = value.lower()
            if normalized in {"true", "yes", "1", "adjusted"}:
                return "adjusted"
            if normalized in {"false", "no", "0", "raw", "unadjusted"}:
                return "unadjusted"
    return "unknown"


def _coerce_time(values: pd.Series) -> pd.Series:
    if pd.api.types.is_datetime64_any_dtype(values):
        return pd.to_datetime(values, utc=True, errors="coerce")

    numeric = pd.to_numeric(values, errors="coerce")
    numeric_count = int(numeric.notna().sum())
    if numeric_count > 0 and numeric_count >= max(1, int(len(values) * 0.8)):
        median_value = float(numeric.dropna().median())
        unit = "ms" if median_value > 10_000_000_000 else "s"
        return pd.to_datetime(numeric, unit=unit, utc=True, errors="coerce")

    return pd.to_datetime(values, utc=True, errors="coerce")


def normalize_historical_dataset(
    symbol: str,
    provider: str,
    market: str,
    timeframe: str,
    raw_dataset: Any,
    metadata: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    metadata = metadata or {}
    issues: List[Dict[str, Any]] = []
    warnings: List[Dict[str, Any]] = []
    diagnostics: Dict[str, Any] = {"input_rows": 0, "dropped_rows": 0, "duplicate_rows_removed": 0}

    frame = _raw_to_frame(raw_dataset)
    diagnostics["input_rows"] = int(len(frame))
    diagnostics["input_columns"] = [str(column) for column in frame.columns]

    if frame.empty and len(frame.columns) == 0:
        return {
            "symbol": symbol,
            "provider": provider,
            "normalized_dataset": [],
            "row_count": 0,
            "timezone": "unknown",
            "currency": metadata.get("currency"),
            "adjusted_status": _metadata_adjusted_status(metadata),
            "normalization_status": "FAILED",
            "issues": [_issue("EMPTY_DATASET", "Raw dataset is empty.")],
            "warnings": warnings,
            "raw_diagnostics": diagnostics,
        }

    if "time" not in frame.columns:
        issues.append(_issue("MISSING_TIME_COLUMN", "Raw dataset has no date/time column."))

    missing_columns = [column for column in REQUIRED_COLUMNS if column not in frame.columns]
    if missing_columns:
        issues.append(_issue("MISSING_OHLCV_COLUMNS", "Raw dataset is missing OHLCV columns.", {"missing_columns": missing_columns}))

    if issues:
        return {
            "symbol": symbol,
            "provider": provider,
            "normalized_dataset": [],
            "row_count": 0,
            "timezone": "unknown",
            "currency": metadata.get("currency"),
            "adjusted_status": _metadata_adjusted_status(metadata),
            "normalization_status": "FAILED",
            "issues": issues,
            "warnings": warnings,
            "raw_diagnostics": diagnostics,
        }

    timezone = str(metadata.get("timezone") or metadata.get("tz") or "unknown")
    raw_time = frame["time"]
    if not pd.api.types.is_datetime64_any_dtype(raw_time) and raw_time.map(lambda value: isinstance(value, str) and not any(token in value for token in ("Z", "+", "-"))).any():
        warnings.append(_issue("TIMEZONE_UNKNOWN", "Provider did not declare timezone for at least one timestamp."))

    frame["time"] = _coerce_time(raw_time)
    if frame["time"].isna().any():
        bad = int(frame["time"].isna().sum())
        warnings.append(_issue("INVALID_TIMESTAMPS_DROPPED", "Rows with invalid timestamps were dropped.", {"rows": bad}))
        diagnostics["dropped_rows"] += bad
        frame = frame.dropna(subset=["time"])

    for column in REQUIRED_COLUMNS + (["adj_close"] if "adj_close" in frame.columns else []):
        frame[column] = pd.to_numeric(frame[column], errors="coerce")

    null_required = frame[REQUIRED_COLUMNS].isna().any(axis=1)
    if bool(null_required.any()):
        count = int(null_required.sum())
        warnings.append(_issue("NULL_OHLCV_DROPPED", "Rows with null OHLCV values were dropped.", {"rows": count}))
        diagnostics["dropped_rows"] += count
        frame = frame[~null_required]

    non_positive_prices = (frame[PRICE_COLUMNS] <= 0).any(axis=1)
    if bool(non_positive_prices.any()):
        count = int(non_positive_prices.sum())
        issues.append(_issue("NON_POSITIVE_PRICES_DROPPED", "Rows with zero or negative prices were rejected.", {"rows": count}))
        diagnostics["dropped_rows"] += count
        frame = frame[~non_positive_prices]

    zero_volume_count = int((frame["volume"].fillna(0) <= 0).sum())
    if zero_volume_count:
        warnings.append(_issue("ZERO_OR_NULL_VOLUME", "Dataset contains zero or null volume.", {"rows": zero_volume_count}))

    before_dedupe = len(frame)
    frame = frame.drop_duplicates(subset=["time", "open", "high", "low", "close", "volume"])
    diagnostics["duplicate_rows_removed"] = before_dedupe - len(frame)
    if diagnostics["duplicate_rows_removed"]:
        warnings.append(_issue("DUPLICATE_ROWS_REMOVED", "Exact duplicate candles were removed.", {"rows": diagnostics["duplicate_rows_removed"]}))

    frame = frame.sort_values("time")
    adjusted_status = _metadata_adjusted_status(metadata)
    if "adj_close" in frame.columns and adjusted_status == "unknown":
        warnings.append(_issue("ADJUSTMENT_AMBIGUOUS", "adj_close exists but OHLC adjustment status is not declared."))

    normalized_dataset: List[Dict[str, Any]] = []
    for row in frame.to_dict(orient="records"):
        timestamp = pd.Timestamp(row["time"])
        normalized_dataset.append(
            {
                "time": int(timestamp.timestamp()),
                "open": float(row["open"]),
                "high": float(row["high"]),
                "low": float(row["low"]),
                "close": float(row["close"]),
                "volume": float(row["volume"]),
            }
        )

    diagnostics["output_rows"] = len(normalized_dataset)
    if not normalized_dataset:
        issues.append(_issue("NO_NORMALIZED_ROWS", "No rows remained after normalization."))

    status = "FAILED" if not normalized_dataset else "WARNING" if issues or warnings else "OK"
    return {
        "symbol": symbol,
        "provider": provider,
        "normalized_dataset": normalized_dataset,
        "row_count": len(normalized_dataset),
        "timezone": timezone,
        "currency": metadata.get("currency"),
        "adjusted_status": adjusted_status,
        "normalization_status": status,
        "issues": issues,
        "warnings": warnings,
        "raw_diagnostics": diagnostics,
    }
